Enable Codecoverage Upload (Jacoco/Codecov)

hazel-sudz opened the pull request:
- [x] verify that it is actually running
- [x] codecov.yml file probably good idea
- [x] clean up for review
- [x] verify that codecov seems reasonable 

basically moving changes over from https://github.com/twitter/scalding/pull/1922 based on https://github.com/twitter/scalding/issues/1977

hazel-sudz committed:
(squash) draft implementation

another miscopy typo

typo in dirname

feed cobertura.xml into codecov instead of scoverage.xml

specify files due to autodetect failure

use coverageAggregate instead of coverageReport

load default codecov behaviour into codecov.yml

typo

draft impl

hazel-sudz commented on .github/workflows/CI.yml:
```diff
@@ -130,3 +130,12 @@ jobs:
           ${{ matrix.script }}
+
+
+      - uses: actions/checkout@master
+      - uses: codecov/codecov-action@v2
+        with:
+          # files: ./coverage1.xml,./coverage2.xml # optional
+          name: codecov
+          fail_ci_if_error: true
+          verbose: true
```
the documentation assures that no credentials are needed using the provided action. I'm not sure how the uploaded works and if it can auto-detect the generated coverage files

hazel-sudz commented on scripts/run_test.sh:
```diff
@@ -5,40 +5,30 @@ BASE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )"/.. && pwd )"
 TST_EXIT_CODE=$?
 
 echo "Running mima checks ... "
-echo "time ./sbt ++$TRAVIS_SCALA_VERSION $(withCmd mimaReportBinaryIssues)"
```
we already have the bash flags that print out commands before they are run so this isn't really needed

hazel-sudz commented on scripts/run_test.sh:
```diff
@@ -5,40 +5,30 @@ BASE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )"/.. && pwd )"
-echo "all done"
 
-exit $(( $TST_EXIT_CODE || $MIMA_EXIT_CODE || $COMPILE_DOC_EXIT_CODE ))
+echo "all test checks done"
```
we already have bash flags to fail-fast so we don't explicitly need to collect exit codes

hazel-sudz committed:
add changes to codecov.yml

hazel-sudz committed:
(squash) don't upload to codecov when there is nothing to upload

johnynek commented:
Coverage seems quite low. Are we sure it is aggregating all the projects correctly? Many subprojects wind up testing the others.

hazel-sudz commented:
> Coverage seems quite low. Are we sure it is aggregating all the projects correctly? Many subprojects wind up testing the others.

there's currently one part in the CI that is failing. It should be a bit higher because of it. I'm not 100% sure why it's failing just get a "step failed" in cascading. I suspect it's due to very high memory usage due to coverage. I can currently reproduce the failure locally when switching coverage on/off before running the test. I'm playing with memory settings but currently not much success. See https://github.com/scoverage/sbt-scoverage/issues/50

hazel-sudz commented:
> > Coverage seems quite low. Are we sure it is aggregating all the projects correctly? Many subprojects wind up testing the others.
> 
> there's currently one part in the CI that is failing. It should be a bit higher because of it. I'm not 100% sure why it's failing just get a "step failed" in cascading. I suspect it's due to very high memory usage due to coverage. I can currently reproduce the failure locally when switching coverage on/off before running the test. I'm playing with memory settings but currently not much success. See [scoverage/sbt-scoverage#50](https://github.com/scoverage/sbt-scoverage/issues/50)

~~after tuning the JVM opts, jacking the stack frame size to 64M (crazy!) seemed to let the failing test pass. It looks like others have had similar issues: https://github.com/scoverage/sbt-scoverage/issues/276 https://github.com/scoverage/sbt-scoverage/issues/422~~

edit: I though I had coverage enabled for above but I did not so I still have no resolution for sbt-scoverage

johnynek commented:
huh:

https://github.com/twitter/scalding/runs/5664208563?check_suite_focus=true#step:7:1367

seems one of the jobs fails (in both 2.11 and 2.12) with coverage... I wonder if somehow the instrumentation is making something non-serializable...

hazel-sudz commented:
> huh:
> 
> https://github.com/twitter/scalding/runs/5664208563?check_suite_focus=true#step:7:1367
> 
> seems one of the jobs fails (in both 2.11 and 2.12) with coverage... I wonder if somehow the instrumentation is making something non-serializable...

that's completely possible. I've tested https://www.scala-sbt.org/sbt-jacoco/index.html as an alternative on the failing test and it seems to work. I'm going to try it in CI. 

The downside of jacoco is that it knows nothing about scala (targets bytecode) so it might be slightly less accurate.

hazel-sudz committed:
(squash) use jacoco instead of scoverage

autodetect on codecov uploader seems to work

new bash features not supported on CI

try using jacoco

hazel-sudz commented on scripts/run_test.sh:
```diff
@@ -4,41 +4,27 @@
-export JVM_OPTS="-XX:+CMSClassUnloadingEnabled -XX:+UseConcMarkSweepGC -XX:ReservedCodeCacheSize=96m -XX:+TieredCompilation -XX:MaxPermSize=256m -Xms256m -Xmx512m -Xss2m"
-
-
-INNER_JAVA_OPTS="set javaOptions += \"-Dlog4j.configuration=file://$TRAVIS_BUILD_DIR/project/travis-log4j.properties\""
```
I see no references to this variable while searching through the codebase. I don't think it's used anymore

hazel-sudz committed:
update README with codecov link

hazel-sudz committed:
try adding some retry to the codecov uploader

hazel-sudz commented:
@johnynek does 40% sound reasonable? It's recording 0% for maple (there is no test folder) and 0% for scalding-avro (no test folder) and 0% for scalding-thrift-macros (there are some tests but maybe it can't trace macros well). 

I think this is similar to what scoverage was reporting. 

Report is here: https://codecov.io/gh/twitter/scalding/tree/5db2df2a50b4a92c0fbf934a5dc13b06229624e9
![Screen Shot 2022-03-24 at 12 28 30 PM](https://user-images.githubusercontent.com/52898838/159995196-aee5da5d-134f-40ec-8df7-4503e2a03475.png)


the reference from 2 years ago was 68%: https://codecov.io/gh/twitter/scalding/tree/f1e9670175047724bf23727ab17f807303281150

![Screen Shot 2022-03-24 at 12 34 38 PM](https://user-images.githubusercontent.com/52898838/159996181-fc3cf881-c525-4a11-b0b4-c16858c8dd78.png)

johnynek commented:
I don't see how it could have fallen so much, but maybe.

I was looking through, it marks a lot of lines as partially covered, e.g.:

https://app.codecov.io/gh/twitter/scalding/blob/5db2df2a50b4a92c0fbf934a5dc13b06229624e9/scalding-core/src/main/scala/com/twitter/scalding/typed/memory_backend/MemoryBackend.scala

are those being counted?

Anyway, this is an improvement, and we can iterate. Thanks for spending the time on this. We can later see if we can find examples of tests that definitely exercise code, but it doesn't see and add an issue at that time.

johnynek commented on README.md:
```diff
@@ -1,7 +1,7 @@
 
 [![Build status](https://github.com/twitter/scalding/actions/workflows/CI.yml/badge.svg?branch=develop)](https://github.com/twitter/scalding/actions)
-[![Coverage Status](https://coveralls.io/repos/twitter/scalding/badge.png?branch=develop)](https://coveralls.io/r/twitter/scalding?branch=develop)
+[![Coverage Status](https://img.shields.io/codecov/c/github/twitter/scalding/develop.svg?maxAge=3600)](https://codecov.io/github/twitter/scalding)
```
max age is 1h? seems like maybe that should be longer no?

hazel-sudz replied:
I'm fine with whatever. Although github also does their own caching so it's probably a bit irrelevant

hazel-sudz commented:
> I don't see how it could have fallen so much, but maybe.
> 
> I was looking through, it marks a lot of lines as partially covered, e.g.:
> 
> https://app.codecov.io/gh/twitter/scalding/blob/5db2df2a50b4a92c0fbf934a5dc13b06229624e9/scalding-core/src/main/scala/com/twitter/scalding/typed/memory_backend/MemoryBackend.scala
> 
> are those being counted?
> 
> Anyway, this is an improvement, and we can iterate. Thanks for spending the time on this. We can later see if we can find examples of tests that definitely exercise code, but it doesn't see and add an issue at that time.

my guess would be that the partial lines are the ones that can throw exceptions. Ie. they are being executed but not all paths covered

johnynek commented:
Thanks!
